CRM/Contribute/Form/Contribution.php

```diff
@@ -1756,7 +1756,18 @@ protected function submit($submittedValues, $action, $pledgePaymentID) {
 
         // to get 'from email id' for send receipt
         $this->fromEmailId = $formValues['from_email_address'];
-        if (CRM_Contribute_Form_AdditionalInfo::emailReceipt($this, $formValues)) {
+        // $tplParams are parameters which are tested to be in the template (CRMContributeFormContributionTest)
+        // and deliberately synchronised with the variables available from the
+        // front end. In time it is hoped they will replace $formValues which is not
+        // clearly defined and predictable.
```

Comment:
I've been doing this a lot in the code, grandfathering in a variable that is clearly defined in order to eventually ditch these nebulous undefined variables.